src/gui/src/layoutViewer.cpp

```diff
@@ -1071,82 +1071,106 @@ void LayoutViewer::drawInstanceShapes(dbTechLayer* layer,
+
+  // core cell text should be 10% of cell height
+  static const float size_target = 0.1;
+  // text should not fill more than 90% of the instance height or width
+  static const float size_limit = 0.9;
+  static const float rotation_limit = 0.85; // slightly lower to prevent oscillating rotations when zooming
+
+  // limit non-core text to 1/2.0 (50%) of cell height or width
+  static const float non_core_scale_limit = 2.0;
 
-  // aim for 1/4 the width of master, 1/20 the height of master,
-  // but no more than 50x bigger than all other text
-  static const int width_scalar = 4;
-  static const int height_scalar = 20;
-  static const int max_scalar = 50;
-  // text should not fill more than 90% of the instance width
-  static const double width_limit = 0.9;
+  // minimum pixel height for text
+  static const int minimum_text_height = 10;
+
+  const float font_core_scale_height = size_target*pixels_per_dbu_;
+  const float font_core_scale_width = size_limit*pixels_per_dbu_;
 
   painter->setTransform(QTransform());
+  painter->setFont(text_font);
   for (auto inst : insts) {
     dbMaster* master = inst->getMaster();
-    if (master->getHeight() < minimum_height) {
+    int master_height = master->getHeight();
+    int master_width = master->getHeight();
+
+    if (inst->getMaster()->isCore() && master_height < minimum_size) {
+      // if core cell, just check master height
+      continue;
+    } else if (master_width < minimum_size || master_height < minimum_size) {
       continue;
     }
 
-    // setup the instance's transform
-    dbTransform inst_xfm;
-    inst->getTransform(inst_xfm);
-
-    // get bbox
-    Rect master_box;
-    master->getPlacementBoundary(master_box);
-    inst_xfm.apply(master_box);
+    Rect instance_box;
+    inst->getBBox()->getBox(instance_box);
 
-    // draw text
     QString name = inst->getName().c_str();
+    QRectF instance_bbox_in_px = dbuToScreen(instance_box);
 
-    QFont inst_font = initial_font;
-    inst_font.setPixelSize(font_size);
-    QRect text_bbox = QFontMetrics(inst_font).boundingRect(name);
-    int inst_width = master_box.dx();
-    int inst_height = master_box.dy();
-    int text_bbox_width = width_scalar*text_bbox.width();
-    int text_bbox_height = height_scalar*text_bbox.height();
-    if (text_bbox_width < inst_width || text_bbox_height < inst_height) {
-      // text will be tiny, so update pixel size
-      int scale_font_width = std::floor(static_cast<double>(inst_width) / text_bbox_width);
-      int scale_font_height = std::floor(static_cast<double>(inst_height) / text_bbox_height);
-      int scale_font = std::min(max_scalar, std::max(scale_font_width, scale_font_height));
-
-      inst_font.setPixelSize(scale_font * font_size);
-      text_bbox = QFontMetrics(inst_font).boundingRect(name);
+    QRectF text_bounding_box = font_metrics.boundingRect(name);
+
+    if (text_bounding_box.height() < minimum_text_height) {
+      continue;
```

Comment:
I think you should check font_metrics.ascent() < minimum_text_height and do that before the instance loop.